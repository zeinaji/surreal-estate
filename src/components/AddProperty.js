import React, { useState } from "react";
import "../styles/AddProperty.css";

import addProperty from "../requests/add-property";
import Alert from "./Alert";

const AddProperty = () => {
  const initialState = {
    fields: {
      title: "",
      city: "Manchester",
      type: "Flat",
      bedrooms: "",
      bathrooms: "",
      price: "",
      email: "",
    },
    alert: {
      message: "",
      isSuccess: false,
    },
  };

  const [fields, setFields] = useState(initialState.fields);
  const [alert, setAlert] = useState(initialState.alert);

  const handleAddProperty = async (event) => {
    event.preventDefault();
    const status = await addProperty(fields);

    if (status === 201) {
      setAlert({ message: "This has been successful", isSuccess: true });
    } else {
      setAlert({ message: "There has been an error", isSuccess: false });
    }
  };

  const handleFieldChange = (event) => {
    const { name, value } = event.target;
    setFields({ ...fields, [name]: value });
  };
  return (
    <div className="add-property">
      {alert.message && (
        <Alert message={alert.message} success={alert.isSuccess} />
      )}
      <form onSubmit={handleAddProperty} noValidate>
        <label>
          Title:
          <input
            id-="title"
            type="text"
            name="title"
            placeholder="Title"
            required
            value={fields.title}
            onChange={handleFieldChange}
          />
        </label>

        <label>
          Type:
          <select
            className="select-css"
            id="type"
            name="type"
            value={fields.type}
            onChange={handleFieldChange}
          >
            <option>Flat</option>
            <option>Detached</option>
            <option>Semi-Detached</option>
            <option>Terraced</option>
            <option>End of Terrace</option>
            <option>Cottage</option>
            <option>Bungalow</option>
          </select>
        </label>

        <label>
          Bedrooms:
          <input
            id-="bedrooms"
            type="number"
            name="bedrooms"
            placeholder="Bedrooms number"
            required
            value={fields.bedrooms}
            onChange={handleFieldChange}
          />
        </label>

        <label>
          Bathrooms:
          <input
            id-="bathrooms"
            name="bathrooms"
            type="number"
            placeholder="Bathrooms number"
            required
            value={fields.bathrooms}
            onChange={handleFieldChange}
          />
        </label>

        <label>
          Price: Â£
          <input
            id-="price"
            name="price"
            type="number"
            placeholder="Price"
            required
            min="0.01"
            step="0.01"
            value={fields.price}
            onChange={handleFieldChange}
          />
        </label>

        <label>
          City
          <select
            className="select-css"
            id="city"
            name="city"
            value={fields.city}
            onChange={handleFieldChange}
          >
            <option>Manchester</option>
            <option>Leeds</option>
            <option>Sheffield</option>
            <option>Liverpool</option>
          </select>
        </label>

        <label>
          Email:
          <input
            id-="email"
            name="email"
            type="email"
            placeholder="example@hotmail.com"
            required
            value={fields.email}
            onChange={handleFieldChange}
          />
        </label>
        <button type="submit">Add</button>
      </form>
    </div>
  );
};

export default AddProperty;
