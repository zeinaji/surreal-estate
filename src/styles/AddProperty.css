@import url("https://fonts.googleapis.com/css2?family=Nunito&display=swap");

.add-property {
  /* padding: 10px;*/
  padding-top: 10px;
  width: 98.5vw;
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: var(--primary-color);
  color: #fff;
  height: 85vh;
  position: relative;
  margin: 10px;
  margin-top: -15px;
}

.add-property::before {
  content: "";
  background: url("../pastel.jpg") no-repeat center center/cover;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: -1;
}

.add-form {
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: rgb(139, 192, 185, 0.8);
  padding: 20px;
}

input::placeholder {
  font-family: "Nunito", sans-serif;
  font-weight: 400;
  padding-left: 4px;
}

.select-css {
  font-family: "Nunito", sans-serif;
  font-weight: 400;
  /* color: rgb(83, 80, 80);*/
  padding-left: 4px;
  height: 25px;
  border-radius: 9px;
  width: 300px;
  display: inline-block;
  float: right;
  outline: none;
}

.select-css::-ms-expand {
  display: none;
}

.label {
  display: inline-block;
  width: 395px;
  text-align: left;
  font-family: "Nunito", sans-serif;
  font-weight: 600;
  color: rgb(83, 80, 80);
  margin-bottom: 10px;
}

.add-form > label > input {
  border-radius: 9px;
  border: 1px solid rgb(83, 80, 80);
  height: 25px;
  width: 300px;
  display: inline-block;
  float: right;
  outline: none;
  padding: 5px;
}

.add-button {
  width: 70px;
  height: 30px;
  margin-top: 10px;
  border-radius: 15px;
  font-family: "Nunito", sans-serif;
  font-weight: 600;
  color: rgb(83, 80, 80);
  border: 1px solid rgb(83, 80, 80);

  background-color: white;
  transition: opacity 0.5s ease in;
  cursor: pointer;
  outline: none;
}

.add-button:hover {
  opacity: 0.7;
}

@media (max-width: 480px) {
  .add-property {
    width: 100vw;
    margin-left: 0;
    margin-right: 0;
  }
  .add-form > label > input,
  .select-css {
    width: 200px;
  }

  .label {
    width: 295px;
  }
}

@media (max-width: 750px) {
  .add-property {
    width: 100vw;
    margin-left: 0;
    margin-right: 0;
  }
}
