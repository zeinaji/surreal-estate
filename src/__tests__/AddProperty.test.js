import React from "react";
import { render, fireEvent, wait, act } from "@testing-library/react";
import AddProperty from "../components/AddProperty";

//requests
import addProperty from "../requests/add-property";
jest.mock("../requests/add-property");

describe("AddProperty", () => {
  let title;
  let type;
  let bedrooms;
  let bathrooms;
  let price;
  let city;
  let email;
  let add;

  beforeEach(() => {
    const { getByText, getByPlaceholderText, getByDisplayValue } = render(
      <AddProperty />
    );

    title = getByPlaceholderText("Title");
    type = getByDisplayValue("Flat");
    bedrooms = getByPlaceholderText("Bedrooms number");
    bathrooms = getByPlaceholderText("Bathrooms number");
    price = getByPlaceholderText("Price");
    city = getByDisplayValue("Manchester");
    email = getByPlaceholderText("example@hotmail.com");
    add = getByText("Add");

    addProperty.mockResolvedValue({ status: 201 });
  });
  xit("renders correctly", () => {
    const { asFragment } = render(<AddProperty />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders all input fields and button", () => {
    expect(title).toHaveAttribute("id-", "title");
    expect(type).toHaveAttribute("id", "type");
    expect(bedrooms).toHaveAttribute("id-", "bedrooms");
    expect(bathrooms).toHaveAttribute("id", "bathrooms");
    expect(price).toHaveAttribute("id-", "price");
    expect(city).toHaveAttribute("id", "city");
    expect(email).toHaveAttribute("id-", "email");
    expect(add).toHaveAttribute("type", "submit");
  });

  xit("catches and posts the user's input", async () => {
    act(() => {
      fireEvent.change(title, {
        target: { value: "City Centre Flat" },
      });
      fireEvent.change(bedrooms, {
        target: { value: 2 },
      });
      fireEvent.change(bathrooms, {
        target: { value: 1 },
      });
      fireEvent.change(price, {
        target: { value: 600 },
      });
      fireEvent.change(city, {
        target: { value: "Sheffield" },
      });
      fireEvent.change(email, {
        target: { value: "exampleemail@example.com" },
      });
    });
    fireEvent.click(add);

    await wait(() => {
      expect(title.value).toBe("City Centre Flat");
      expect(type.value).toBe("Flat");
      expect(bedrooms.value).toBe("2");
      expect(bathrooms.value).toBe("1");
      expect(price.value).toBe("600");
      expect(city.value).toBe("Sheffield");
      expect(email.value).toBe("exampleemail@example.com");

      expect(addProperty).toHaveBeenCalledTimes(1);
      expect(addProperty).toHaveBeenCalledWith({
        title: "City Centre Flat",
        type: "Flat",
        bedrooms: "2",
        bathrooms: "1",
        price: 600,
        city: "Sheffield",
        email: "exampleemail@example.com",
      });
    });
  });

  it("it doesn't render the Alert component before anything is added", () => {
    const { queryByText } = render(<AddProperty />);
    expect(queryByText(/successfully/)).not.toBeInTheDocument();
    expect(queryByText(/error/)).toBeFalsy();
  });

  it("renders the Alert component when the request is successful", async () => {
    const { getByText } = render(<AddProperty />);
    addProperty.mockResolvedValue({
      status: 404,
    });
    fireEvent.click(add);

    await wait(() => expect(getByText(/error/)).toBeInTheDocument());
  });

  it("renders the Alert component when the request is succesful", async () => {
    const { getByText } = render(<AddProperty />);
    addProperty.mockResolvedValue({ status: 201 });
    fireEvent.click(add);

    await wait(() => expect(getByText(/successfully/)).toBeInTheDocument());
  });
});
