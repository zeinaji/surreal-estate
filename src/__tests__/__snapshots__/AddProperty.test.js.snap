// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProperty renders correctly 1`] = `
<DocumentFragment>
  <div
    class="add-property"
  >
    <form
      class="add-form"
    >
      <label>
        Title:
        <input
          id="title"
          name="title"
          placeholder="Title"
          required=""
          value=""
        />
      </label>
      <label>
        Type:
        <select
          class="select-css"
          id="type"
          name="type"
        >
          <option>
            Flat
          </option>
          <option>
            Detached
          </option>
          <option>
            Semi-Detached
          </option>
          <option>
            Terraced
          </option>
          <option>
            End of Terrace
          </option>
          <option>
            Cottage
          </option>
          <option>
            Bungalow
          </option>
        </select>
      </label>
      <label>
        Bedrooms:
        <input
          id="bedrooms"
          name="bedrooms"
          placeholder="Bedrooms number"
          required=""
          value="0"
        />
      </label>
      <label>
        Bathrooms:
        <input
          id="bathrooms"
          name="bathrooms"
          placeholder="Bathrooms number"
          required=""
          value="0"
        />
      </label>
      <label>
        Price: Â£
        <input
          id="price"
          name="price"
          placeholder="Price"
          type="number"
          value="0"
        />
      </label>
      <label>
        City
        <select
          class="select-css"
          id="city"
          name="city"
        >
          <option>
            Manchester
          </option>
          <option>
            Leeds
          </option>
          <option>
            Sheffield
          </option>
          <option>
            Liverpool
          </option>
        </select>
      </label>
      <label>
        Email:
        <input
          id="email"
          name="email"
          placeholder="example@hotmail.com"
          required=""
          type="email"
          value=""
        />
      </label>
      <button
        class="add-button"
        type="submit"
      >
        Add
      </button>
    </form>
  </div>
</DocumentFragment>
`;
